### ISSUE TITLE
Underline-style h2 after list interpreted as hr

### ISSUE THREAD

**Issue by christhekeele** [pragdave/earmark#168]
The `---` form for h2 headers are treated as horizontal rules when immediately following a list.

For example:

```elixir
"""
foo

bar
---

baz
""" |> Earmark.as_html! |> IO.puts
```

correctly becomes:

```html
<p>foo</p>
<h2>bar</h2>
<p>baz</p>
```

However:

```elixir
"""
* foo

bar
---

baz
""" |> Earmark.as_html! |> IO.puts
```

incorrectly becomes:

```html
<ul>
<li>foo
</li>
</ul>
<p>bar</p>
<hr class="thin"/>
<p>baz</p>
```

**Comment by christhekeele** [comment]
It looks like even this suffers from it:

```elixir
Earmark.as_html! "foo\n___\n"
#=> "<p>foo</p>\n<hr class=\"medium\"/>\n"
```

**Comment by RobertDober** [comment]
@christhekeele thank you for your report, both cases are clear bugs, however it seems that your second example is fixed in 1.2.4, but, alas, not the first case.

**Comment by christhekeele** [comment]
💟
